### PR TITLE
net/hdlc: initial hdlc layer

### PR DESCRIPTION
**Pull request by maxvankessel**
I've created with help from @jia200x  a HDLC layer (netdev layer),  it's part of the PPP stack.
It should be layered on top of a PPPoS netdev layer, which handles incoming bytes and the ACCM map.
But it can also be layered on top of a generic modem driver (netdev). 

This layers knows it's a netdev_layer.

I would like some feedback on the buffers is the hdlc_t, I don't like it this way.

### CODE REVIEW

File: sys/include/net/hdlc/hdr.h [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief HDLC types
+ */
+typedef enum {
+    HDLC_TYPE_RECEIVE_READY     = 0,        /**< ready receive type */
```
Comment (kaspar030) [inline sys/include/net/hdlc/hdr.h]: I find the shifts quite confusing. Would you mind using hex values?

File: sys/include/net/hdlc/hdr.h [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+typedef enum {
+    HDLC_FRAME_TYPE_INFORMATION = 0, /**< represents i(nformation) frame */
+    HDLC_FRAME_TYPE_SUPERVISORY = 1, /**< represents s(upervisory) frame*/
+    HDLC_FRAME_TYPE_UNNUMBERED = 3,  /**< represents u(numbered) frame */
+} hdlc_frame_type_t;
+
+/**
+ * @brief HDLC Supervisory Frames
+ */
+typedef struct __attribute__((packed)) {
```
Comment (kaspar030) [inline sys/include/net/hdlc/hdr.h]: Please don't use bitfields for wire formats. The actual order is implementation defined and thus cannot be relied upon.

File: sys/include/net/hdlc.h [hunk 3]
```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (C) 2018 Max van Kessel
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for
+ * more details.
+ */
+
+/**
+ * @defgroup    net_hdlc
```
Comment (smlng) [inline sys/include/net/hdlc.h]: new group needs a title

File: sys/include/net/hdlc/fcs.h [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+#define NET_HDLC_FCS_H
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/* Table for fast calculation of fcs16 */
```
Comment (smlng) [inline sys/include/net/hdlc/fcs.h]: missing documentation, i.e., add

```
/**
 * @brief <something descriptive>
 */
```

File: sys/include/net/hdlc/hdr.h [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    hdlc_type_t type : 2;
+    uint8_t poll_final : 1;
+    uint8_t sequence_no : 3;
+} hdlc_control_s_frame_t;
+
+/**
+ * @brief HDLC Information Frames
+ */
+typedef struct __attribute__((packed)) {
+    uint8_t id : 1;
```
Comment (smlng) [inline sys/include/net/hdlc/hdr.h]: struct members need docu, too

File: sys/include/net/hdlc/hdr.h [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    HDLC_FRAME_TYPE_INFORMATION = 0, /**< represents i(nformation) frame */
+    HDLC_FRAME_TYPE_SUPERVISORY = 1, /**< represents s(upervisory) frame*/
+    HDLC_FRAME_TYPE_UNNUMBERED = 3,  /**< represents u(numbered) frame */
+} hdlc_frame_type_t;
+
+/**
+ * @brief HDLC Supervisory Frames
+ */
+typedef struct __attribute__((packed)) {
+    uint8_t id : 2;
```
Comment (smlng) [inline sys/include/net/hdlc/hdr.h]: please add docu for struct members

File: sys/include/net/hdlc/hdr.h [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    uint8_t send_sequence_no : 3;
+    uint8_t poll_final : 1;
+    uint8_t sequence_no : 3;
+} hdlc_control_i_frame_t;
+
+/**
+ * @brief HDLC Unnumbered Frames
+ */
+typedef struct __attribute__((packed)) {
+    uint8_t id : 2;
```
Comment (smlng) [inline sys/include/net/hdlc/hdr.h]: dito, docu for struct members

File: sys/include/net/hdlc/hdr.h [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+ *  +--------------------------------+
+ *  +----------+----------+-----------------
+ *  |   FCS    |   Flag   | Inter-frame Fill
+ *  |16/32 bits| 01111110 | or next Address
+ *  +----------+----------+-----------------
+ *
+ */
+typedef struct __attribute__((packed)) {
+    uint8_t address;                /**< Address field oh HDLC header */
+    union{
```
Comment (smlng) [inline sys/include/net/hdlc/hdr.h]: add space `union {`

File: sys/include/net/hdlc/hdr.h [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        hdlc_control_s_frame_t s;
+        hdlc_control_i_frame_t i;
+        hdlc_control_u_frame_t u;
+        uint8_t frame;              /**< Control field of HDLC header */
+    }control;
```
Comment (smlng) [inline sys/include/net/hdlc/hdr.h]: add space `} control;` and docu on this member

File: sys/include/net/hdlc/fcs.h [hunk 10]
```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2016 José Ignacio Alamos <jialamos@uc.cl>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for
+ * more details.
+ */
+
+/**
+ * @defgroup    net_hdlc_fcs    FCS calculation
```
Comment (smlng) [inline sys/include/net/hdlc/fcs.h]: not sure about having sub-groups for fas and hdr, could use `net_hdlc` as well (IMHO) but its up to you,

File: sys/include/net/hdlc.h [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+ * @brief   Device descriptor for HDLC
+ *
+ * @extends netdev_t
+ */
+typedef struct {
+    netdev_t netdev;                /**< parent class */
+    tsrb_t inbuf;                   /**< RX buffer */
+    uint8_t rxmem[HDLC_BUFSIZE];    /**< memory used by RX buffer */
+    uint8_t txmem[HDLC_BUFSIZE];    /**< memory used by TX buffer */
+
```
Comment (smlng) [inline sys/include/net/hdlc.h]: stray empty line? remove please

File: sys/include/net/hdlc.h [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+ */
+#ifndef HDLC_BUFSIZE
+#define HDLC_BUFSIZE (128)
+#endif
+
+/**
+ * @brief Default timeout for retransmission of sequence flag
+ */
+#ifndef HDLC_MAX_IDLE_TIME_MS
+#define HDLC_MAX_IDLE_TIME_MS   (100 * US_PER_MS)
```
Comment (smlng) [inline sys/include/net/hdlc.h]: should include `xtimer.h` above when using this conversion macro

File: sys/include/net/hdlc/fcs.h [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+ * @ingroup     net_hdlc
+ * @brief       FCS calculation routine
+ *
+ * @{
+ *
+ * @file
+ * @brief       Routine for DCS calculation
```
Comment (smlng) [inline sys/include/net/hdlc/fcs.h]: typo: /DCS/FCS/ ?

File: sys/include/net/hdlc/hdr.h (hunk 1 above)
Comment (smlng) [reply sys/include/net/hdlc/hdr.h]: I agree, this looks a bit weird here

**Comment by maxvankessel** [comment]
I've updated the source accordingly.

**Comment by jia200x** [comment]
ping @maxvankessel

**Comment by jia200x** [comment]
I like the approach of keeping HDLC and PPPoS in different layers as you did! I can ACK the fundamentals

**Comment by jia200x** [comment]
Some feedback from Vera++:

```
sys/net/link_layer/hdlc/hdlc.c:284: too many consecutive empty lines
sys/include/net/hdlc.h:65: comma should be followed by whitespace
sys/include/net/hdlc/fcs.h:72: line is longer than 100 characters
sys/net/link_layer/hdlc/hdlc.c:284: too many consecutive empty lines 
sys/include/net/hdlc.h:65: comma should be followed by whitespace
sys/include/net/hdlc/fcs.h:72: line is longer than 100 characters
sys/net/link_layer/hdlc/hdlc.c:284: too many consecutive empty lines
```

File: sys/include/net/hdlc/hdr.h [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+    HDLC_FRAME_TYPE_UNNUMBERED  = 3, /**< represents u(numbered) frame */
+} hdlc_frame_type_t;
+
+/**
+ * @brief Data type to represent an HDLC header.
+ *
+ *  +----------+----------+----------+
+ *  |   Flag   | Address  | Control  |
+ *  | 01111110 | 11111111 | 00000011 |
+ *  +----------+----------+----------+
```
Comment (miri64) [inline sys/include/net/hdlc/hdr.h]: I'm unsure how this translates to the header defined below.

Comment (miri64) [inline sys/include/net/hdlc/hdr.h]: For it to be rendered correctly as pixelart in the rendered doc, this needs to be indented some spaces (not sure how many, please check yourself with `make doc`)

Comment (jia200x) [reply sys/include/net/hdlc/hdr.h]: The flag is not a part of the header but a delimiter of the packet

**Comment by miri64** [comment]
Ping @maxvankessel

**Comment by jia200x** [comment]
ping @maxvankessel

**Comment by maxvankessel** [comment]
I wil have a look at this. Is it still useful to implement this layer after all this time?

**Comment by miri64** [comment]
> I will have a look at this. Is it still useful to implement this layer after all this time?

We still want PPP support, so yes :-).

**Comment by maxvankessel** [comment]
Okay, give me some time to get back into it, it's been a while

File: sys/include/net/hdlc/fcs.h [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+    0x39c3, 0x284a, 0x1ad1, 0x0b58, 0x7fe7, 0x6e6e, 0x5cf5, 0x4d7c,
+    0xc60c, 0xd785, 0xe51e, 0xf497, 0x8028, 0x91a1, 0xa33a, 0xb2b3,
+    0x4a44, 0x5bcd, 0x6956, 0x78df, 0x0c60, 0x1de9, 0x2f72, 0x3efb,
+    0xd68d, 0xc704, 0xf59f, 0xe416, 0x90a9, 0x8120, 0xb3bb, 0xa232,
+    0x5ac5, 0x4b4c, 0x79d7, 0x685e, 0x1ce1, 0x0d68, 0x3ff3, 0x2e7a,
+    0xe70e, 0xf687, 0xc41c, 0xd595, 0xa12a, 0xb0a3, 0x8238, 0x93b1,
+    0x6b46, 0x7acf, 0x4854, 0x59dd, 0x2d62, 0x3ceb, 0x0e70, 0x1ff9,
+    0xf78f, 0xe606, 0xd49d, 0xc514, 0xb1ab, 0xa022, 0x92b9, 0x8330,
+    0x7bc7, 0x6a4e, 0x58d5, 0x495c, 0x3de3, 0x2c6a, 0x1ef1, 0x0f78
+};
```
Comment (benpicco) [inline sys/include/net/hdlc/fcs.h:67]: There is already [`sys/checksum/crc16_ccitt.c`](https://doc.riot-os.org/group__sys__checksum__crc16__ccitt.html)

**Comment by jia200x** [comment]
hi @maxvankessel 

I would really like to move forward with this. Do you plan to continue working on this anytime?
Otherwise, I can offer to take over this PR (keeping your authorship, of course)

**Comment by maxvankessel** [comment]
Hello @jia200x,
Im not planning to revive this pr. You're free to use it.

**Pull request closed without merging**